### PR TITLE
Fix error message for assigning to non-null properties.

### PR DESCRIPTION
**Pull request by NightlyNexus**
instead of falling down to "Required property 'a' missing at $"

### CODE REVIEW

File: kotlin/src/test/java/com/squareup/moshi/kotlin/KotlinJsonAdapterTest.kt [hunk 1]
```diff
@@ -152,6 +152,20 @@ class KotlinJsonAdapterTest {
...
+    val moshi = Moshi.Builder().add(object {
+      @FromJson fun fromJson(string: String): String? = null
+    }).add(KotlinJsonAdapterFactory()).build()
+    val jsonAdapter = moshi.adapter(HasNonNullConstructorParameter::class.java)
+
+    try {
+      jsonAdapter.fromJson("{\"a\":\"hello\"}")
+      fail()
+    } catch (expected: JsonDataException) {
+      assertThat(expected).hasMessage("Non-null value 'a' was null at \$")
```
Comment (NightlyNexus) [inline kotlin/src/test/java/com/squareup/moshi/kotlin/KotlinJsonAdapterTest.kt]: existing behavior with incorrect path (same as nonNullConstructorParameterCalledWithNullFailsWithJsonDataException) #510.

File: kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt [hunk 2]
```diff
@@ -383,6 +383,20 @@ class GeneratedAdaptersTest {
...
+    val moshi = Moshi.Builder().add(object {
+      @FromJson fun fromJson(string: String): String? = null
+    }).build()
+    val jsonAdapter = moshi.adapter(HasNonNullConstructorParameter::class.java)
+
+    try {
+      jsonAdapter.fromJson("{\"a\":\"hello\"}")
+      fail()
+    } catch (expected: JsonDataException) {
+      assertThat(expected).hasMessage("Non-null value 'a' was null at \$.a")
```
Comment (NightlyNexus) [inline kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt]: note the different message from the null literal. `"Unexpected null at \$.a"`

**Comment by swankjesse** [comment]
I like the behavior but I don’t like the impact on the generated code. Is there a way to get good behavior with less generated code?

**Comment by swankjesse** [comment]
In particular, I think it’s worthwhile making different tradeoffs with generated vs. reflective code. We don’t want to generate more code than we have to, even if it makes the errors a little more awkward in corner cases. (And in practice I don’t think many adapters take non-null values and return null.)

**Comment by NightlyNexus** [comment]
Yeah, I'll take another look and close out if there's not a way to do it with less code. Agree it's probably a rare case.

**Comment by swankjesse** [comment]
I’m gonna merge as-is ’cause it‘s a net win. We can figure out whether we want to shrink the generated code later.

**Comment by NightlyNexus** [comment]
cool, i'm pretty sure this can be improved.

**Pull request merged**